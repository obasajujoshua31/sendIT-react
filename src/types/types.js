export const signUpUser = "SIGN_UP_USER";
export const signInUser = "SIGN_IN_USER";
export const createOrder = "CREATE_ORDER";
export const createOrderSuccess = "CREATE_ORDER_SUCCESS";
export const createOrderFailure = "CREATE_ORDER_FAILURE";
export const viewOrders = "VIEW_ORDERS";
export const viewOrder = "VIEW_ORDER";
export const updateParcel = "UPDATE_PARCELS";
export const updateLocation = "UPDATE_PARCEL_LOCATION";
export const cancelOrder = "CANCEL_ORDER";
export const updateParcelStatus = "UPDATE_PARCEL_STATUS";
export const signInSuccess = "SIGN_IN_SUCCESS";
export const signInFailure = "SIGN_IN_FAILURE";
export const signUpSuccess = "SIGN_UP_SUCCESS";
export const signUpFailure = "SIGN_UP_FAILURE";
export const isLoggedOut = "LOGGED_OUT";
export const isAdmin = "IS_ADMIN";
