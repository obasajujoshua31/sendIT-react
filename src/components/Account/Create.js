import React, { Component } from "react";
const emailRegex = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/;

const isValid = ({ formErrors, ...rest }) => {
  let valid = true;
  Object.values(formErrors).forEach(
    value => value.length > 0 && (valid = false)
  );

  Object.values(rest).forEach(value => value.length === 0 && (valid = false));
  return valid;
};
class CreateAccount extends Component {
  constructor() {
    super();
    this.state = {
      formErrors: {
        firstName: "",
        lastName: "",
        email: "",
        password: "",
        confirmPassword: ""
      },
      firstName: "",
      lastName: "",
      email: "",
      password: "",
      confirmPassword: ""
    };
    this.handleChange = this.handleChange.bind(this);
  }
  handleChange(e) {
    const { name, value } = e.target;
    const { formErrors, password } = this.state;
    switch (name) {
      case "firstName":
        formErrors.firstName =
          value.length > 2 ? "" : "First name cannot be less than 2 characters";
        break;
      case "lastName":
        formErrors.lastName =
          value.length > 2 ? "" : "Last name cannot be less than 2 characters";
        break;
      case "email":
        formErrors.email = emailRegex.test(value) ? "" : "Invalid Email";
        break;
      case "password":
        formErrors.password = value.length > 5 ? "" : "Weak Password";
        break;
      case "confirmPassword":
        formErrors.confirmPassword =
          value === password ? "" : "Password does not match";
    }
    this.setState({ [name]: value }, () => console.log(this.state));
  }
  render() {
    const {
      firstName,
      lastName,
      email,
      password,
      confirmPassword,
      formErrors
    } = this.state;

    return <React.Fragment>
        <div className="account-wrapper">
          <div className="form-wrapper create-wrapper">
            <form onSubmit={this.handleSubmit} noValidate>
              <h2>Create Account</h2>
              <div className="form-inline">
                <div className="form-group">
                  <label htmlFor="firstName">First Name</label>
                  <input className={formErrors.firstName.length > 0 ? "error" : ""} type="text" placeholder="First Name" name="firstName" value={firstName} onChange={this.handleChange} />
                  {formErrors.firstName.length > 0 && <span className="errorMsg">
                      {formErrors.firstName}
                    </span>}
                </div>
                <div className="form-group">
                  <label htmlFor="lastName">Last Name</label>
                  <input className={formErrors.lastName.length > 0 ? "error" : ""} type="text" placeholder="last Name" name="lastName" value={lastName} onChange={this.handleChange} />
                  {formErrors.lastName.length > 0 && <span className="errorMsg">
                  {formErrors.lastName}  </span>}
                </div>
                
                
              </div>
              <div className="form-group">
                <label htmlFor="email">Email</label>
                <input className={formErrors.email.length > 0 ? "error" : ""} type="email" placeholder="yourname@example.com" name="email" value={email} onChange={this.handleChange} />
              {formErrors.email.length > 0 && (<span className='errorMsg'>{formErrors.email}</span>)} 
              </div>
           
              <div className="form-group">
                <label htmlFor="password">Password</label>
                <input className={formErrors.password.length > 0 ? "error" : ""} type="password" placeholder="password" name="password" value={password} onChange={this.handleChange} />
              {formErrors.password.length > 0 && (<span className='errorMsg'>{formErrors.password}</span>)} 
              </div>
            
              <div className="form-group">
                <label htmlFor="confirmPassword">Confirm Password</label>
                <input className={formErrors.confirmPassword.length > 0 ? "error" : ""} type="password" placeholder="confirm password" name="confirmPassword" value={confirmPassword} onChange={this.handleChange} />
              {formErrors.confirmPassword.length > 0 && (<span className='errorMsg'>{formErrors.confirmPassword}</span>)} 
              </div>
              <button type="submit">Register</button>
            </form>
          </div>
        </div>
      </React.Fragment>;
  }
}
export default CreateAccount;
