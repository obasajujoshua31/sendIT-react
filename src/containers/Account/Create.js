import React, { Component } from "react";
import { connect } from "react-redux";
import Spinner from "react-md-spinner";
import { withRouter } from "react-router-dom";
import validateCreateForm, { isValid } from "../../validations/createValidator";
import registerUser from "../../actions/user/signup";
import { signUpSuccess } from "../../types/types";
import { toast } from "react-toastify";

export class CreateAccount extends Component {
  constructor(props) {
    super(props);
    this.state = {
      formErrors: {
        firstName: "",
        lastName: "",
        email: "",
        password: "",
        confirmPassword: ""
      },
      firstName: "",
      lastName: "",
      email: "",
      password: "",
      confirmPassword: "",
      isDeactivated: false,
      isLoading: false
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }
  handleChange(e) {
    const { name, value } = e.target;
    this.setState({ formErrors: validateCreateForm(e, this.state) });
    this.setState({ [name]: value });
    if (!isValid(this.state)) {
      this.setState({ isDeactivated: true });
    } else {
      this.setState({ isDeactivated: false });
    }
  }

  handleSubmit(e) {
    const { firstName, lastName, email, password } = this.state;
    e.preventDefault();
    this.setState({ isLoading: true });
    if (!isValid(this.state)) {
      return this.setState({ isLoading: false });
    } else {
      this.props.signUp({ firstName, lastName, email, password });
    }
  }

  componentDidUpdate(prevProps) {
    if (prevProps !== this.props) {
      if (this.props.signUpStatus === signUpSuccess) {
        this.props.history.push("/view-orders/");
        toast.success("Welcome to SendIT");
      }
    }
  }

  render() {
    if (this.state.isLoading) {
      return (
        <div className="container">
          <div className="table-container">
            <div className="container-not-found">
              <Spinner />
            </div>
          </div>
        </div>
      );
    }

    const {
      firstName,
      lastName,
      email,
      password,
      confirmPassword,
      formErrors,
      isDeactivated
    } = this.state;
    return (
      <React.Fragment>
        <div className="account-wrapper">
          <div className="form-wrapper create-wrapper">
            <form onSubmit={this.handleSubmit} noValidate>
              <h2>Create Account</h2>

              <div className="form-oneline">
                <div className="form-group">
                  <label htmlFor="firstName">First Name</label>
                  <input
                    className={formErrors.firstName.length > 0 ? "error" : ""}
                    type="text"
                    placeholder="First Name"
                    name="firstName"
                    value={firstName}
                    onChange={this.handleChange}
                    required
                  />
                  {formErrors.firstName.length > 0 && (
                    <span className="errorMsg">{formErrors.firstName}</span>
                  )}
                </div>
                <div className="form-group">
                  <label htmlFor="lastName">Last Name</label>
                  <input
                    className={formErrors.lastName.length > 0 ? "error" : ""}
                    type="text"
                    placeholder="last Name"
                    name="lastName"
                    value={lastName}
                    onChange={this.handleChange}
                    required
                  />
                  {formErrors.lastName.length > 0 && (
                    <span className="errorMsg">{formErrors.lastName} </span>
                  )}
                </div>
              </div>
              <div className="form-group">
                <label htmlFor="email">Email</label>
                <input
                  className={
                    formErrors.email.length > 0 ||
                    this.props.errorMsg.length > 0
                      ? "error"
                      : ""
                  }
                  type="email"
                  placeholder="yourname@example.com"
                  name="email"
                  value={email}
                  onChange={this.handleChange}
                  required
                />
                {this.props.errorMsg.length > 0 && (
                  <span className="errorMsg">{this.props.errorMsg}</span>
                )}
                {formErrors.email.length > 0 && (
                  <span className="errorMsg">{formErrors.email}</span>
                )}
              </div>

              <div className="form-group">
                <label htmlFor="password">Password</label>
                <input
                  className={formErrors.password.length > 0 ? "error" : ""}
                  type="password"
                  placeholder="password"
                  name="password"
                  value={password}
                  onChange={this.handleChange}
                  required
                />
                {formErrors.password.length > 0 && (
                  <span className="errorMsg">{formErrors.password}</span>
                )}
              </div>

              <div className="form-group">
                <label htmlFor="confirmPassword">Confirm Password</label>
                <input
                  className={
                    formErrors.confirmPassword.length > 0 ? "error" : ""
                  }
                  type="password"
                  placeholder="confirm password"
                  name="confirmPassword"
                  value={confirmPassword}
                  onChange={this.handleChange}
                  required
                />
                {formErrors.confirmPassword.length > 0 && (
                  <span className="errorMsg">{formErrors.confirmPassword}</span>
                )}
              </div>
              <button hidden={isDeactivated} type="submit">
                Register
              </button>
            </form>
          </div>
        </div>
      </React.Fragment>
    );
  }
}
export const mapStateToProps = ({ users }) => {
  return {
    signUpStatus: users.signUpStatus,
    errorMsg: users.errorMsg
  };
};

export const mapDispatchToProps = dispatch => {
  return {
    signUp: user => dispatch(registerUser(user))
  };
};
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(CreateAccount));
