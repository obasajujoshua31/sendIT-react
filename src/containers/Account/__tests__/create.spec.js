import React from "react";
import { shallow, mount } from "enzyme";
import sinon from "sinon";
import { CreateAccount, mapDispatchToProps, mapStateToProps } from "../Create";
import { signUpSuccess } from "../../../types/types";

const props = {
  signUp: jest.fn(),
  history: {
    push: jest.fn()
  },
  errorMsg: "",
  signUpStatus: ""
};
const mockUser = {
  firstName: "Monkey",
  lastName: "Gorilla",
  email: "monkey@gorilla.com",
  password: "animal",
  confirmPassword: "animal"
};

const mockStore = {
  users: {
    signUpStatus: "Success",
    errorMsg: null
  }
};

const wrapper = shallow(<CreateAccount {...props} />);

describe(" <CreateAccount> ", () => {
  it("should render CreateAccount Component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should contain 8 <div> tags", () => {
    expect(wrapper.find("div").length).toEqual(8);
  });
  it("should contain 1 <form> tag", () => {
    expect(wrapper.find("form").length).toEqual(1);
  });
  it("should contain 1 <h2> tag", () => {
    expect(wrapper.find("h2").length).toEqual(1);
  });
  it("should contain 5 <label> tags", () => {
    expect(wrapper.find("label").length).toEqual(5);
  });
  it("should contain 5 <input> tags", () => {
    expect(wrapper.find("input").length).toEqual(5);
  });
  it("should contain 1 <button> tag", () => {
    expect(wrapper.find("button").length).toEqual(1);
  });
  it("should set the firstName field to test", () => {
    wrapper
      .find("input")
      .at(0)
      .simulate("change", {
        target: {
          value: "test",
          name: "firstName"
        }
      });
    expect(wrapper.instance().state.firstName).toEqual("test");
  });

  // it("should set isDeactivated to false", () => {
  //   wrapper
  //     .find("input")
  //     .at(0)
  //     .simulate("change", {
  //       target: {
  //         value: "",
  //         name: "firstName"
  //       }
  //     });
  //   expect(wrapper.instance().state.firstName).toEqual("");
  //   // expect(wrapper.instance().state.isDeactivated).toEqual(true)
  // });

  it("should call dispatch action", () => {
    const dispatchSpy = sinon.spy();
    const dispatchResult = mapDispatchToProps(dispatchSpy);
    dispatchResult.signUp(mockUser);
    expect(dispatchSpy.callCount).toEqual(1);
  });

  it("should map state to Props", () => {
    const mapStateResult = mapStateToProps(mockStore);
    expect(mapStateResult.signUpStatus).toEqual("Success");
  });

  it("should call history.push", () => {
    const addedProps = { ...props, signUpStatus: signUpSuccess };
    const alteredWrapper = shallow(<CreateAccount {...addedProps} />);
    expect(alteredWrapper.instance().props.history.push).toHaveBeenCalled();
  });

  it("should call submit", () => {
    let prevented = false;
    const formElement = wrapper.find("form");
    formElement.simulate("submit", {
      preventDefault: () => {
        prevented = true;
      }
    });
    expect(prevented).toEqual(true);
  });

  it("should call props.signup", () => {
    let prevented = false;
    wrapper.setState({
      firstName: mockUser.firstName,
      lastName: mockUser.lastName,
      email: mockUser.email,
      password: mockUser.password,
      confirmPassword: mockUser.confirmPassword
    });
    const formElement = wrapper.find("form");
    formElement.simulate("submit", {
      preventDefault: () => {
        prevented = true;
      }
    });
    expect(wrapper.instance().props.signUp).toHaveBeenCalled();
  });
});
