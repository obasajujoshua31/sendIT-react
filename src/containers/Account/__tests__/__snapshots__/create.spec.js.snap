// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` <CreateAccount>  should render CreateAccount Component 1`] = `
<Fragment>
  <div
    className="account-wrapper"
  >
    <div
      className="form-wrapper create-wrapper"
    >
      <form
        noValidate={true}
        onSubmit={[Function]}
      >
        <h2>
          Create Account
        </h2>
        <div
          className="form-oneline"
        >
          <div
            className="form-group"
          >
            <label
              htmlFor="firstName"
            >
              First Name
            </label>
            <input
              className=""
              name="firstName"
              onChange={[Function]}
              placeholder="First Name"
              required={true}
              type="text"
              value=""
            />
          </div>
          <div
            className="form-group"
          >
            <label
              htmlFor="lastName"
            >
              Last Name
            </label>
            <input
              className=""
              name="lastName"
              onChange={[Function]}
              placeholder="last Name"
              required={true}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="email"
          >
            Email
          </label>
          <input
            className=""
            name="email"
            onChange={[Function]}
            placeholder="yourname@example.com"
            required={true}
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="password"
          >
            Password
          </label>
          <input
            className=""
            name="password"
            onChange={[Function]}
            placeholder="password"
            required={true}
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="confirmPassword"
          >
            Confirm Password
          </label>
          <input
            className=""
            name="confirmPassword"
            onChange={[Function]}
            placeholder="confirm password"
            required={true}
            type="password"
            value=""
          />
        </div>
        <button
          hidden={false}
          type="submit"
        >
          Register
        </button>
      </form>
    </div>
  </div>
</Fragment>
`;
